frontend/web/src/app/components/digital-print/digital-print-edit/digital-print-edit.component.ts

```diff
@@ -0,0 +1,194 @@
+      (digitalPrint) => {
+        this.digitalPrint = digitalPrint;
+        this.populate();
+      },
+    );
+  }
+
+  private constructMagicPairingFormGroup(){
+    /* load attr with default data from product instance*/
+    this.magicPairingsFieldsForm = [];
+    Object.getOwnPropertyNames(this.digitalPrint.magicsPartPairing).forEach((id) => {
+      this.magicPairingsFieldsForm.push(this.formBuilder.group({
+        magicsId: [id,
+          Validators.compose([Validators.required])],
+        digitalPart: [this.digitalPrint.magicsPartPairing[id] ? this.digitalPrint.magicsPartPairing[id] : '',
+          Validators.compose([Validators.required])],
+      }));
+    });
+    console.log(this.magicPairingsFieldsForm);
+  }
+
+  deleteMagicPairing(option) {
+    this.magicPairingsFieldsForm.splice(option, 1);
+  }
+
+  addMagicPairing() {
+    const fields = {
+      magicsId: ['',
+        Validators.compose([Validators.required])],
+      digitalPart: ['',
+        Validators.compose([Validators.required])],
+    };
+    this.magicPairingsFieldsForm.push(this.formBuilder.group(fields));
+  }
+
+  public pairingValid(): boolean {
+    let valid = true;
+    this.magicPairingsFieldsForm.forEach((formGroup) => {
+      if (!formGroup.valid) {
+        valid = false;
+        return;
+      }
+    });
+    return valid;
+  }
+
+  /* populate data */
+  private populate() {
+    this.constructForms();
+  }
+
+  /* init forms */
+  private constructForms() {
+    const fields = {
+      magicsPath: [this.digitalPrint && this.digitalPrint.magicsPath ? this.digitalPrint.magicsPath : '',
+        Validators.compose([Validators.required])],
+    };
+
+    this.requiredFieldsForm = this.formBuilder.group(fields);
+    this.constructMagicPairingFormGroup();
+    this.loaded = true;
+  }
+
+  /* save product instance */
+  save() {
+    /* convert relevant fields */
+
+    const id = this.digitalPrint.id;
+    const digitalPrint: DigitalPrint = new DigitalPrint(this.requiredFieldsForm.value);
+    this.creating ? delete digitalPrint['id'] : digitalPrint.id = id;
+
+    const pairings = {};
+    this.magicPairingsFieldsForm.forEach((formGroup) => {
+      pairings[formGroup.get('magicsId').value] = parseInt(formGroup.get('digitalPart').value);
+    });
+    digitalPrint.magicsPartPairing = pairings;
+
+    console.log(digitalPrint);
+    console.log(JSON.stringify(digitalPrint));
+
+    if (this.creating) { // a new product
+      this.digitalPrintService.createDigitalPrint(digitalPrint).subscribe(
+        (data) => {
+          if (this.nav) {
+            this.back();
+          }
+          this.changed.emit(data);
+        }, (error) => {
+          console.log(error);
+          this.errorService.showAlert(error.verobose_message_header, error.verbose_message);
+        },
+      );
+    } else {
+      this.digitalPrintService.updateDigitalPrint(digitalPrint).subscribe(
+        (data) => {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Update this function so that its implementation is not identical to the one on line 158. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=typescript%3AS4144)